### PR TITLE
[NewOptimizer] Hook up printing, print linetable

### LINKED ISSUE #27149: [NewOptimizer] (Legacy) IR does not show line number information anymore
```
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: https://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.7.0-DEV.4839 (2018-04-10 06:39 UTC)
 _/ |\__'_|_|_|\__'_|  |  Commit 000f243d5d* (38 days old master)
|__/                   |  x86_64-pc-linux-gnu

julia> @code_warntype sin(1)
Variables:
  x::Int64

Body:
  begin
      # meta: location float.jl float 266
      # meta: location float.jl Type 251
      # meta: location float.jl Type 57
      Core.SSAValue(4) = (Base.sitofp)(Float64, x::Int64)::Float64
      # meta: pop locations (3)
      Core.SSAValue(1) = $(Expr(:invoke, MethodInstance for sin(::Float64), :(Base.Math.sin), Core.SSAValue(4)))::Float64
      return Core.SSAValue(1)
  end::Float64
```

```
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: https://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.7.0-DEV.5127 (2018-05-18 07:05 UTC)
 _/ |\__'_|_|_|\__'_|  |  master/ecf24255ac* (fork: 4086 commits, 303 days)
|__/                   |  x86_64-pc-linux-gnu

julia> @code_warntype sin(1)
Variables:
  x::Int64

Body:
  begin
      (Base.sitofp)(Float64, x::Int64)::Float64
      $(Expr(:invoke, MethodInstance for sin(::Float64), :(Base.Math.sin), Core.SSAValue(1)))::Float64
      return Core.SSAValue(2)
  end::Float64
```

### PR DESCRIPTION
**Pull request by Keno**
This hooks up printing for the new IR and teaches it to print the
linetable. In printing a linetable, I tried to strike a balance
between verboseness (I think the old printing and our printing of LLVM IR
is significantly too verbose). There's a description of the idea in
show.jl.

Screenshot:
screen shot 2018-05-19 at 5 38 38 pm

Fixes #27149

### CODE REVIEW

**Comment by Keno** [comment]
By request now includes an optional mode where the line table is printed more verbosely:
screen shot 2018-05-20 at 9 03 55 pm

File: base/show.jl [hunk 1]
```diff
@@ -630,16 +630,32 @@ function show(io::IO, l::Core.MethodInstance)
...
     lambda_io = IOContext(io, :SOURCEINFO => src)
     if src.slotnames !== nothing
         lambda_io = IOContext(lambda_io, :SOURCE_SLOTNAMES => sourceinfo_slotnames(src))
     end
-    body = Expr(:body)
-    body.args = src.code
-    show(lambda_io, body)
+    if src.codelocs !== nothing
+        println(io)
+        ir = Core.Compiler.inflate_ir(src)
```
Comment (vtjnash) [inline base/show.jl]: Base typically should not depend directly on `Core.Compiler`, since we don't guarantee that it exists and/or won't get removed or replaced at runtime

File: stdlib/InteractiveUtils/src/codeview.jl [hunk 2]
```diff
@@ -32,33 +20,71 @@ function code_warntype(io::IO, f, @nospecialize(t))
...
+problematic for performance, so the results need to be used judiciously.
+In particular, unions containing either [`missing`](@ref) or [`nothing`](@ref) are displayed in yellow, since
+these are often intentional.
+If the `verbose_linetable` keyword is set, the linetable will be printed
+in verbose mode, showing all available information (rather than applying
+the usual heuristics).
+See [`@code_warntype`](@ref man-code-warntype) for more information.
+"""
+function code_warntype(io::IO, f, @nospecialize(t); verbose_linetable=false)
+    io = IOContext(io, :verbose_linetable => verbose_linetable)
```
Comment (vtjnash) [inline stdlib/InteractiveUtils/src/codeview.jl]: Have this exist as both a context argument and keyword argument seems unfortunate. Maybe just drop the keyword argument? Although, I've slowly come to agree with Jeff's dislike of having user customization at all, over just picking the "right" answer as a default.

**Comment by MikeInnes** [comment]
Note: I guess this is still intended to be used with Keno's fork of Revise. As-is an IRCode object will not display:
```julia
julia> Zygote.@code_ir gcd(10,5)
Error showing value of type Core.Compiler.IRCode:
ERROR: MethodError: no method matching get(::IOContext{REPL.Terminals.TTYTerminal}, ::Symbol, ::Bool)
You may have intended to import Base.get
```
But this should be as simple to fix as forwarding the Core.Compiler methods to Base, I hope.

File: base/compiler/ssair/show.jl [hunk 3]
```diff
@@ -37,69 +39,266 @@ function print_node(io::IO, idx, stmt, used, maxsize; color = true, print_typ=tr
...
+        if print_typ && stmt.typ !== Any
+            Base.print(io, "::$(stmt.typ)")
+        end
+    elseif isexpr(stmt, :invoke)
+        print(io, "invoke ")
+        linfo = stmt.args[1]
+        print_ssa(io, stmt.args[2], argnames)
+        Base.print(io, "(")
+        print_arg((typ, arg)) = sprint(io->(print_ssa(io, arg, argnames); print(io, "::$(typ)")))
+        Base.print(io, join(map(print_arg, zip(linfo.specTypes.parameters[2:end], stmt.args[3:end])), ", "))
```
Comment (vtjnash) [inline base/compiler/ssair/show.jl]: This code is very hard to parse. Why are you stringifying everything multiple times? At least skip the obvious copy:
```julia
Base.join(io, map(print_arg, zip(linfo.specTypes.parameters[2:end], stmt.args[3:end])), ", ")
```

`linfo.specTypes.parameters[:]` - this quantity doesn't always exist. use `fieldtype(linfo.specTypes, i)`.

As I've said before, using `zip` on `args` (in Inference) is usually a very bad idea (very, very slow, and poisons the dispatch, Tuple, and Type caches for later with too many entries). There's some helper functions in Core.Compiler (such as `_any`, `contains_is`, and `anymap`) which help ensure Inference compiles fast and small which should be used instead (and add more if you need them).

Comment (vtjnash) [reply base/compiler/ssair/show.jl]: (actually, don't use `fieldtype`, since doing this approximate transform at all will convey incorrect information about the invoke site)
